src/Workspaces/Core/Portable/FindSymbols/FindReferences/BaseTypeFinder.cs

```diff
@@ -0,0 +1,103 @@
+{
+    internal static partial class BaseTypeFinder
+    {
+        public static async Task<ImmutableArray<SymbolAndProjectId>> FindBaseTypesAndInterfacesAsync(
+            INamedTypeSymbol type, Project project, CancellationToken cancellationToken)
+        {
+            var typesAndInterfaces = FindBaseTypesAndInterfaces(type);
+            return await ConvertToSymbolAndProjectIdsAsync(typesAndInterfaces.CastArray<ISymbol>(), project, cancellationToken).ConfigureAwait(false);
+        }
+
+        public static async Task<ImmutableArray<SymbolAndProjectId>> FindOverriddenAndImplementedMembersAsync(
+            ISymbol symbol, Project project, CancellationToken cancellationToken = default)
+        {
+            var solution = project.Solution;
+            var baseClassesAndInterfaces = FindBaseTypesAndInterfaces(symbol.ContainingType);
+            var results = ArrayBuilder<SymbolAndProjectId>.GetInstance();
+
+            // These are implicit interface implementaitons not matching by name such as void I.M();
+            results.AddRange(
+                await ConvertToSymbolAndProjectIdsAsync(
+                    symbol.ExplicitInterfaceImplementations(),
+                    project,
+                    cancellationToken).ConfigureAwait(false));
+
+            foreach (var type in baseClassesAndInterfaces)
+            {
+                foreach (var m in type.GetMembers(symbol.Name))
+                {
+                    var sourceMember = await SymbolFinder.FindSourceDefinitionAsync(
+                        SymbolAndProjectId.Create(m, project.Id),
+                        solution,
+                        cancellationToken).ConfigureAwait(false);
+
+                    if (sourceMember.Symbol != null)
+                    {
+                        // These are explicit interface imlementations matching by name.
+                        if (type?.TypeKind == TypeKind.Interface)
+                        {
+                            if (symbol.ContainingType?.TypeKind == TypeKind.Class)
+                            {
+                                var implementation = symbol.ContainingType.FindImplementations(sourceMember.Symbol, solution.Workspace);
+
+                                if (implementation != null &&
+                                    SymbolEquivalenceComparer.Instance.Equals(implementation.OriginalDefinition, symbol.OriginalDefinition))
+                                {
+                                    results.Add(sourceMember);
+                                }
+                            }
+                        }
+                        else if (SymbolFinder.IsOverride(solution, symbol, sourceMember.Symbol, cancellationToken))
+                        {
+                            results.Add(sourceMember);
+                        }
+                    }
+                }
+            }
+
+            return results.ToImmutableAndFree();
```

Comment:
so i feel not great about thsi approach.  it feels very duplicative of logic we have elsewhere.  I'd prefer if we just used the normal FindRefs infrastructure.  It already does this cascading logic for you.  All we'd need to do is say that we want the symbol cascading but *not* the references.  

Note: this is also necessary for correctness.  If you're trying to find the interface members implemented by a member, you can't do it by walking the declared interfaces of your base types.

Reply:
I tried to follow [https://github.com/dotnet/roslyn/blob/c420bead8f60f6582903b8d8676227bbd178f67b/src/EditorFeatures/Core/FindUsages/FindUsagesHelpers.cs#L68](https://github.com/dotnet/roslyn/blob/c420bead8f60f6582903b8d8676227bbd178f67b/src/EditorFeatures/Core/FindUsages/FindUsagesHelpers.cs#L68)
which leads to 
[https://github.com/dotnet/roslyn/blob/c420bead8f60f6582903b8d8676227bbd178f67b/src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Hierarchy.cs](https://github.com/dotnet/roslyn/blob/c420bead8f60f6582903b8d8676227bbd178f67b/src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Hierarchy.cs)

Here, I found a similar idea but for the hierarchy goes down.

Likely, I should make more researches how to reuse existing code.

---
In reply to: [283974069](https://github.com/dotnet/roslyn/pull/35704#discussion_r283974069) [](ancestors = 283974069)

Reply:
What is wrong with the approach for "find the interface members implemented by a member"? Get all interfaces implemented by the current type and then try to match members in those interfaces for the current one.

interface I { void [|M|](); }
class C : I  { public virtual void [|M|]() {} }
class D : C  { public override void $$M() {} }

---
In reply to: [283974069](https://github.com/dotnet/roslyn/pull/35704#discussion_r283974069) [](ancestors = 283974069)

Reply:
```c#
interface I { void M(); }
class C { public void [|M|](); }
```

Should this return I.M?  Based on that (and just looking at C's interfaces), no.  However:

```c#
interface I { void M(); }
class C { public void [|M|](); }

// other project
class D : C, I {
}
```

Now C.M becomes an impl of I.M.

Reply:
note: the above is an extremely esoteric case, and it's likely ok to not catch it.  However, using the actual FindRefs *would* find this.

Reply:
The example is very interesting. Thank you, @cyrusnajmabadi! 

However, I am still not sure that I.M should be considered as a base for C.M. If considering "go to base" as a reverse action to "go to implementation", it could be considered as a base. However, defining "bases" as overridden or implemented methods, I.M should not be a base.

So, it is a corner case, and I do not have a 100% assurance for each of approaches.

---
In reply to: [284045520](https://github.com/dotnet/roslyn/pull/35704#discussion_r284045520) [](ancestors = 284045520)

Reply:
I reviewed FindRefs. It seems it provides much more than the current functionality.

1. It seems complicated to extend FindRefs for the purpose.

2. We could try to extract DetermineCascadedSymbolsAsync from AbstractMethodOrPropertyOrEventSymbolReferenceFinder into a static helper, and then call it from both FindRefs and GoToBase. We also should tune DetermineCascadedSymbolsAsync for GoToBase purposes. So, in my understanding complications of the code would bring more potential issues than benefits.

Please let me know your thought. Thank you!

---
In reply to: [284070734](https://github.com/dotnet/roslyn/pull/35704#discussion_r284070734) [](ancestors = 284070734)